INTMDB-584: [CFN-PI] Create EncryptionAtRest CDK L3 (#325)
mongodb/mongodbatlas-cloudformation-resources: cdk/l3-cdk-resources/encryption-at-rest-express/README.md

const app = new cdk.App();
const stack = new cdk.Stack(app, 'atlas-encryption-at-rest-express', {
  env: { region: process.env.CDK_DEFAULT_REGION, account: process.env.CDK_DEFAULT_ACCOUNT },
});

const PROJECT_ID = stack.node.tryGetContext('MONGODB_ATLAS_PROJECT_ID') || process.env.MONGODB_ATLAS_PROJECT_ID;
const ROLE_ID = stack.node.tryGetContext('MONGODB_ATLAS_ROLE_ID') || process.env.MONGODB_ATLAS_ROLE_ID;
const MASTER_KEY_ID = stack.node.tryGetContext('MONGODB_ATLAS_MASTER_KEY_ID') || process.env.MONGODB_ATLAS_MASTER_KEY_ID;
const DB_USER_PASSWORD = stack.node.tryGetContext('MONGODB_ATLAS_DB_USER_PASSWORD') || process.env.MONGODB_ATLAS_DB_USER_PASSWORD;

new AtlasEncryptionAtRestExpress(this, 'EncryptionAtRestExpress', {
      cluster: {
        name: "ClusterName",
      },

      accessList: {
        accessList: [{
          ipAddress: "192.0.0.1"
        }]
      },

      encryptionAtRest: {
        customerMasterKeyId: MONGODB_ATLAS_MASTER_KEY_ID,
        roleId: ROLE_ID,
      },

      databaseUser: {
        password: DB_USER_PASSWORD
      },

      projectId: PROJECT_ID,
    })
  }
```
The constructor will deploy the following resources using default values:

> - Cluster

Reviewer: Should we also include MongoDB version 5.0 here?

Author: yes, thanks!

Author: updated

# @mongodbatlas-awscdk/encryption-at-rest-express

The official [MongoDB Atlas](https://www.mongodb.com/) AWS CDK resource for Node.js.

> > This construct uses MongoDB [L1 construct](https://constructs.dev/search?q=&offset=0&tags=mongodb-published) and data structures for the [AWS CloudFormation Registry] L3 type.

Reviewer 2: ```suggestion
> This L3 construct uses MongoDB Atlas [L1 constructs](https://constructs.dev/search?q=&offset=0&tags=mongodb-published) and data structures (including Cluster, Database User, IP Access List, and Encryption at Rest).
```

[L1 construct]: https://docs.aws.amazon.com/cdk/latest/guide/constructs.html
[AWS CloudFormation Registry]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry.html

## Description

> The resource lets you create a cluster, dbuser, add an entry to the atlas access list and set encryption-at-rest in your MongoDB Atlas Project.

Reviewer 2: ```suggestion
The resource lets you create an Atlas Cluster, DB User, add an entry to the IP Access List and set encryption-at-rest in your MongoDB Atlas Project.
```

  --type RESOURCE \
  --execution-role-arn ROLE-ARN  

aws cloudformation activate-type \
  --type-name MongoDB::Atlas::ProjectIpAccessList \
  --publisher-id bb989456c78c398a858fef18f2ca1bfc1fbba082 \
  --type RESOURCE \
  --execution-role-arn ROLE-ARN  
    
```

Alternatively:

```sh
aws cloudformation activate-type \
  --public-type-arn arn:aws:cloudformation:us-east-1::type/resource/bb989456c78c398a858fef18f2ca1bfc1fbba082/MongoDB-Atlas-Cluster \
  --execution-role-arn ROLE-ARN

aws cloudformation activate-type \
  --public-type-arn arn:aws:cloudformation:us-east-1::type/resource/bb989456c78c398a858fef18f2ca1bfc1fbba082/MongoDB-Atlas-EncryptionAtRest \
  --execution-role-arn ROLE-ARN  

aws cloudformation activate-type \
  --public-type-arn arn:aws:cloudformation:us-east-1::type/resource/bb989456c78c398a858fef18f2ca1bfc1fbba082/MongoDB-Atlas-DatabaseUser \
  --execution-role-arn ROLE-ARN

aws cloudformation activate-type \
  --public-type-arn arn:aws:cloudformation:us-east-1::type/resource/bb989456c78c398a858fef18f2ca1bfc1fbba082/MongoDB-Atlas-ProjectIpAccessList \
  --execution-role-arn ROLE-ARN
    
```

### Impotant Notes

> 1. Atlas encrypts all cluster storage and snapshot volumes, securing all cluster data on disk: a concept known as encryption at rest, by default.

Reviewer 2: ```suggestion
1. This CDK L3 resource allows users to create and edit the Encryption at Rest using Customer Key Management configuration. MongoDB Cloud encrypts all storage by default whether or not you use your own key management.

```

Author: updated

2. Atlas limits this feature to dedicated cluster tiers of M10 and greater. For more information see: https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Encryption-at-Rest-using-Customer-Key-Management

> 3. This construnctor supports only [AWS Key Management Service](https://www.mongodb.com/docs/atlas/security-aws-kms/#security-aws-kms) 

Reviewer 2: ```suggestion
3. This construnctor supports only [AWS Key Management Service](https://www.mongodb.com/docs/atlas/security-aws-kms/#security-aws-kms).
```

4. MongoDB does not support creating and configuring a IAM role via CFN and CDK. **Please follow this guide ([Enable Role-Based Access to Your Encryption Key for a Project](https://www.mongodb.com/docs/atlas/security-aws-kms/#enable-customer-managed-keys-with-aws-kms)) before trying to use this constructor as you must provide** `roleId` **and the** `customerMasterKeyId`.

> ### Minimal configuration to use this app

Reviewer 2: ```suggestion
### Minimal configuration to use this CDK L3 construct 
```

Author: updated

In this example, @mongodbatlas-awscdk/encryption-at-rest-express is used to enable encryption at rest for your MongoDB Atlas Project.

```typescript
import * as cdk from 'aws-cdk-lib';
import { AtlasEncryptionAtRestExpress } from './index';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'atlas-encryption-at-rest-express', {
  env: { region: process.env.CDK_DEFAULT_REGION, account: process.env.CDK_DEFAULT_ACCOUNT },
});

const PROJECT_ID = stack.node.tryGetContext('MONGODB_ATLAS_PROJECT_ID') || process.env.MONGODB_ATLAS_PROJECT_ID;
const ROLE_ID = stack.node.tryGetContext('MONGODB_ATLAS_ROLE_ID') || process.env.MONGODB_ATLAS_ROLE_ID;
const MASTER_KEY_ID = stack.node.tryGetContext('MONGODB_ATLAS_MASTER_KEY_ID') || process.env.MONGODB_ATLAS_MASTER_KEY_ID;

new AtlasEncryptionAtRestExpress(stack, 'atlas-encryption-at-rest-express', {
  projectId: PROJECT_ID,

  encryptionAtRest: {
    roleId: ROLE_ID,
    customerMasterKeyId: MASTER_KEY_ID,
  },
});
```

### Minimal configuration to create a cluster, database User, configure atlas access list and enable encryption at rest

> In this example we are providing the minimal configuration to the constructor to define an Atlas Cluster, DB user, add an entry to the Atlas Access list, and enable encryption at rest for the cluster and project.

Reviewer 2: ```suggestion
In this example we are providing the minimal configuration to the constructor to define an Atlas Cluster, DB User, add an entry to the Atlas Access list, and enable encryption at rest for the cluster and project.
```

Author: updated

import { AtlasEncryptionAtRestExpress } from './index';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'atlas-encryption-at-rest-express', {
  env: { region: process.env.CDK_DEFAULT_REGION, account: process.env.CDK_DEFAULT_ACCOUNT },
});

const PROJECT_ID = stack.node.tryGetContext('MONGODB_ATLAS_PROJECT_ID') || process.env.MONGODB_ATLAS_PROJECT_ID;
const ROLE_ID = stack.node.tryGetContext('MONGODB_ATLAS_ROLE_ID') || process.env.MONGODB_ATLAS_ROLE_ID;
const MASTER_KEY_ID = stack.node.tryGetContext('MONGODB_ATLAS_MASTER_KEY_ID') || process.env.MONGODB_ATLAS_MASTER_KEY_ID;
const DB_USER_PASSWORD = stack.node.tryGetContext('MONGODB_ATLAS_DB_USER_PASSWORD') || process.env.MONGODB_ATLAS_DB_USER_PASSWORD;

new AtlasEncryptionAtRestExpress(this, 'EncryptionAtRestExpress', {
      cluster: {
        name: "ClusterName",
      },

      accessList: {
        accessList: [{
          ipAddress: "192.0.0.1"
        }]
      },

      encryptionAtRest: {
        customerMasterKeyId: MONGODB_ATLAS_MASTER_KEY_ID,
        roleId: ROLE_ID,
      },

      databaseUser: {
        password: DB_USER_PASSWORD
      },

      projectId: PROJECT_ID,
    })
  }
```

> The constructor will deploy the following resources using default values:

Author: customers have to provide the projectId, this constructor won't create the project. We cannot create the project since the user has to Enable Role-Based Access for a Project and we do not support the cloudProviderAccess endpoint via CFN.

new AtlasEncryptionAtRestExpress(this, 'EncryptionAtRestExpress', {
      cluster: {
        name: "Cluster0",
        backupEnabled: false,
        mongoDbMajorVersion: "6.0",
        replicationSpecs: [
          {
            numShards: 1,
            advancedRegionConfigs: [
              {
                regionName: "US_WEAST_2",
                electableSpecs: {
                  instanceSize: "M10",
                  nodeCount: 6,
                  ebsVolumeType: "STANDARD"
                },
                priority: 7
            }
          ]}
        ]
      },

      accessList: {
        accessList: [{
          ipAddress: "192.0.0.1"
        }]
      },

      encryptionAtRest: {
        customerMasterKeyId: MONGODB_ATLAS_MASTER_KEY_ID,
        roleId: ROLE_ID,
      },

      databaseUser: {
        password: DB_USER_PASSWORD
      },

      projectId: PROJECT_ID,
    })
  }
```

You can find more information about activating this type in the [AWS CloudFormation documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry-public.html).

## Feedback

> * Issues related to this generated library should be [reported here](https://github.com/cdklabs/cdk-cloudformation/issues/new?title=Issue+with+%40cdk-cloudformation%2Fmongodb-atlas-cluster+v1.0.0).

Reviewer 2: ```suggestion
* Issues related to this CDK Labs generated library should be [reported here](https://github.com/cdklabs/cdk-cloudformation/issues/new?title=Issue+with+%40cdk-cloudformation%2Fmongodb-atlas-cluster+v1.0.0).
```

Author: Updated

> * Issues related to this library should be reported to the [publisher](https://github.com/mongodb/mongodbatlas-cloudformation-resources/issues).

Author: updated

* Feature requests should be [reported here](https://feedback.mongodb.com/forums/924145-atlas?category_id=392596)

[cdklabs/cdk-cloudformation]: https://github.com/cdklabs/cdk-cloudformation

## License

> Distributed under the Apache-2.0 License.# replace this

Reviewer 2: ```suggestion
Distributed under the Apache-2.0 License.
```

Author: updated
